File: kotlin/src/main/java/RomanToInteger.java
public class RomanToInteger {

    public int romanToInt(String s) {
        int sum = 0;
        int previous = 0;
        int current;
        for (int i = s.length() - 1; i >= 0; i--) {
            current = charToInt(s.charAt(i));
            if (current < previous) {
                sum -= current;
            } else {
                sum += current;
            }
            previous = current;
        }
        return sum;
    }

    private int charToInt(char c) {
        return switch (c) {
            case 'I' -> 1;
            case 'V' -> 5;
            case 'X' -> 10;
            case 'L' -> 50;
            case 'C' -> 100;
            case 'D' -> 500;
            case 'M' -> 1000;
            default -> 0;
        };
    }
}

File: kotlin/src/main/kotlin/RomanToInteger.kt
package org.wyekings

/**
 * https://leetcode.cn/problems/roman-to-integer
 */
class RomanToInteger {

    fun romanToInt(s: String): Int {
        var sum = 0
        var previous = 0
        for (i in s.length - 1 downTo 0) {
            val current = s[i].int
            if (current < previous) {
                sum -= current
            } else {
                sum += current
            }
            previous = current
        }
        return sum
    }

}

private val Char.int: Int
    get() = when (this) {
        'I' -> 1
        'V' -> 5
        'X' -> 10
        'L' -> 50
        'C' -> 100
        'D' -> 500
        'M' -> 1000
        else -> 0
    }
